### PR TITLE
ext_proc: propagate tracing context

### PR DESCRIPTION
**Pull request by cainelli**
Commit Message: ext_proc: propagate tracing context

Additional Description: Propagate tracing headers in bidirectional GRPC streams to enable tracing in external processors. I mostly followed the suggestion [described](https://github.com/envoyproxy/envoy/issues/21119#issuecomment-1487536484) in the issue. Please note this is my first C++ code, I kindly ask to give as many references you can requesting changes, there's probably a bunch.

Risk Level: Low
Testing: Ran the existing tests (`bazel test --test_output=streamed //test/common/grpc/...`) and manually tested tracing with datadog using a few GRPC / HTTP implementations (ext_proc, ext_authz, ratelimit, lua http call). This was [my test setup](https://github.com/cainelli/envoy-grpc) and here is a trace report in datadog with these changes (orange is envoy spans and yellow ones collected in the external processor:

image

Docs Changes: n/a
Release Notes: extp_roc: propagate tracing headers to the external processor.
Platform Specific Features: n/a
Fixes https://github.com/envoyproxy/envoy/issues/21119

(I messed up https://github.com/envoyproxy/envoy/pull/33565)

### CODE REVIEW

**Comment by cainelli** [comment]
Following up on https://github.com/envoyproxy/envoy/pull/33565#issuecomment-2061660995 

/assign @wbpcode

File: contrib/sip_proxy/filters/network/source/tra/tra_impl.cc
Comment (wbpcode) [inline contrib/sip_proxy/filters/network/source/tra/tra_impl.cc:132]: Rather than take the `parent_span` as a new parameter, may be we can move the span related options from `RequestOptions` to the `StreamOptions`?

**Review by wbpcode — COMMENTED** [review]
Thanks for you contribution. This is a huge PR and first comment to start the review.

File: contrib/sip_proxy/filters/network/source/tra/tra_impl.cc
Comment (cainelli) [reply contrib/sip_proxy/filters/network/source/tra/tra_impl.cc:132]: That was also what I [considered](https://github.com/envoyproxy/envoy/pull/33565#issuecomment-2060088670), I was mostly following existing pattern and [suggestion](https://github.com/envoyproxy/envoy/issues/21119#issuecomment-1487536484) but happy to try this approach.

Comment (wbpcode) [reply contrib/sip_proxy/filters/network/source/tra/tra_impl.cc:132]: I prefer the one to set the parent span to streamoptions. could you split it (set the parent span in the StreamOptions) to a separated PR? I think that PR would be very simple and we can let other maintainers to check if that has some unknown drawbacks or not.

Comment (cainelli) [reply contrib/sip_proxy/filters/network/source/tra/tra_impl.cc:132]: I've managed to get it working, but will wait for comments on the StreamOption approach https://github.com/envoyproxy/envoy/pull/33747 before integrating those changes here. One concern is the inconsistency in how parent_span is set in different places; it's unclear which method takes precedence. Curious to hear your thoughts on this.

**Review by cainelli — COMMENTED** [review]
I left out the `GoogleAsyncClientImpl` for now. I can make it another PR once this approach is reviewed.
 https://github.com/envoyproxy/envoy/blob/c15b875d2f03f583d8bbaa3a43325196c69b73de/source/common/grpc/google_async_client_impl.cc#L80

**Comment by adisuissa** [comment]
@wbpcode PTAL

**Comment by adisuissa** [comment]
Please also merge main again.

**Comment by phlax** [comment]
removing my assignment and deps label - i think they were added as a result of merge mash

**Comment by wbpcode** [comment]
Will take a look tomorrow. Thanks for you updating.

File: envoy/http/async_client.h
Comment (wbpcode) [inline envoy/http/async_client.h:383]: Now, because these fields were moved to StreamOptions from RequestOptions, we should handle these fields in the `operator==` of StreamOptions.

File: envoy/http/async_client.h
Comment (wbpcode) [inline envoy/http/async_client.h:448]: Now, I think the `RequestOpions` could be an alias of `StreamOptions` because it has no any own field.

**Review by wbpcode — COMMENTED** [review]
cool. Thanks for you great contribution. LGTM overall except some minor comments.

**Comment by wbpcode** [comment]
cc @tyxia could you take a look again because I know you use this filter heavily.

**Comment by cainelli** [comment]
@wbpcode @tyxia I pushed the requested changes, can you please take a look again?

**Comment by phlax** [comment]
@cainelli wondering if this should have a changelog - not sure, but it seems like a behaviour change

**Comment by cainelli** [comment]
> @cainelli wondering if this should have a changelog - not sure, but it seems like a behaviour change

Makes sense, I added a line in the changelog.

/retest

File: test/extensions/filters/http/ext_proc/client_test.cc
Comment (wbpcode) [inline test/extensions/filters/http/ext_proc/client_test.cc:82]: I think this update minor changed the original test. In the original test, the stream info in the ParentContext is non-null. But now, it's null.

**Review by wbpcode — COMMENTED** [review]
Thanks for your contributions and sorry for the delay. This is LGTM overall with some minor comments.

Will merge this PR after these minor comments are resolved and no objections from code owners.

**Comment by wbpcode** [comment]
/unassign @tyxia
/assign @yanavlasov 

Seems that @tyxia is busy recently, assign this to another code owner.

File: source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc
Comment (cainelli) [reply source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc]: I changed as I was getting the following error after changing RequestOptions to an alias of StreamOptions, removed the `struct` instead.

```
source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc:51:36: error: type alias 'RequestOptions' cannot be referenced with a struct specifier
  struct Envoy::Http::AsyncClient::RequestOptions options =
```

File: test/extensions/filters/http/ext_proc/client_test.cc
Comment (cainelli) [reply test/extensions/filters/http/ext_proc/client_test.cc]: Nice catch! Thank you.

File: source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc
Comment (wbpcode) [reply source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc]: Got it. Actually, you can just remove the leading struct. 🤣

Comment (wbpcode) [reply source/extensions/filters/http/gcp_authn/gcp_authn_impl.cc]: Hah, I see. You have did in that way.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks so much for your contributions and patience.

Defer the merge to tomorrow in case the code owner has additional comments.

**Review by yanavlasov — APPROVED** [review]

**Comment by cainelli** [comment]
> @cainelli would it be possible to add end-to end test with tracing enabled to ext_proc_integration_test? Otherwise it feels like we have tests that validate bits a pieces of the tracing but nothing that makes sure that all expected spans were created.

@yanavlasov I managed to get a basic integration test checking if the tracing headers are being propagated to extproc but not sure if that was what you had in mind. I used the opentracing filter which I assume is not ideal but did not find a test tracing filter. 

I had a hard time to understand how the integration tests works and failed to find a reference how to test spans there. Please let me know if that's sufficient for now or if you have any pointer to another filter doing it.

**Comment by tyxia** [comment]
> > @cainelli would it be possible to add end-to end test with tracing enabled to ext_proc_integration_test? Otherwise it feels like we have tests that validate bits a pieces of the tracing but nothing that makes sure that all expected spans were created.
> 
> @yanavlasov I managed to get a basic integration test checking if the tracing headers are being propagated to extproc but not sure if that was what you had in mind. I used the opentracing filter which I assume is not ideal but did not find a test tracing filter.
> 
> I had a hard time to understand how the integration tests works and failed to find a reference how to test spans there. Please let me know if that's sufficient for now or if you have any pointer to another filter doing it.

@cainelli Will it be possible to add a simple test tracing filter? test/extensions/filters/http/ext_proc/logging_test_filter.cc could be an example though it is for testing logging purpose.

Having an e2e integration test here can not only validate the functionality of this PR but also prevent other people from breaking your feature accidentally.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (tyxia) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:211]: It is not recommended to have cross-filter dependence as it could be disable in people's custom build.

Would it possible to have a simple test filter similar to OpenTelemetryConfig?

https://github.com/envoyproxy/envoy/tree/main/test/integration/filters this folder contains test filters for your reference. 

From a quick scan, OpenTelemetryConfig performs http/gRPC callout. `server_factory_context_filter_config.proto` could be a reference, though it is only gRPC based callout.  
Besides, I am wondering if your feature can be validated by local filter/extension (i.e., no callout) , which might also make your test easier.

**Review by tyxia — COMMENTED** [review]
Thank you for your contribution!

**Comment by cainelli** [comment]
@tyxia I have added a test tracing filter for the integration tests, really apologies for any eventual non-idiomatic/bad pattern code, as mentioned this is my first c++ contribution and I'm trying hard. The places I wasn't so sure and would appreciate any input.

* The overall proposed architecture for tests tracing in integration tests, it feels mock could be useful but I did not find a way to inject and assert during the integration tests. If that is feasible and it makes more sense please let me know how you'd approach it.
* If mock is not doable, is it ok asserting on the destructor? It just felt okish but I'm not sure of side effects with this approach.
* I only added a couple of integration tests so you get the concept but could add more if the overall approach is acceptable.
* I could move this test tracer to `test/integration/filters` if you think it is generic enough for other filters.

File: source/common/grpc/async_client_impl.cc
Comment (tyxia) [inline source/common/grpc/async_client_impl.cc:270]: Should we `setTag` value here before finishSpan? 
This is local close path. IIUC,  similarly  `Canceled` is used in asyncRequest path 
+@wbpcode  who is trace expert , WDYT?

(Also, please see my comment below on how to test this path)

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (tyxia) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:631]: I think `RequestMessageTimeout` test case in this file will hit the path I commented above

Do you want to add `RequestMessageTimeoutWithTracing` test case?

**Review by tyxia — COMMENTED** [review]
Thank you for your contribution. Great work!  
Mostly LGTM, I left a few comments around localClose case.

File: source/common/grpc/async_client_impl.cc
Comment (cainelli) [reply source/common/grpc/async_client_impl.cc:271]: Makes sense. I added the tag and the test as suggested.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
